src/node/internal/process.ts

```diff
@@ -20,13 +20,90 @@ export function nextTick(cb: Function, ...args: unknown[]) {
+          }
+          return true;
+        case Array.prototype:
+          for (let elem of value) {
+            if (!isJsonSerializable(elem)) return false;
+          }
+          return true;
+        default:
+          return false;
+      }
+    }
+
+    default:
+      return false;
+  }
+}
+
+function getInitialEnv() {
+  let env: Record<string, string> = {};
+  for (let [key, value] of Object.entries(utilImpl.getEnvObject())) {
+    // Workers environment variables can have a variety of types, but process.env vars are
+    // strictly strings. We want to convert our workers env into process.env, but allowing
+    // process.env to contain non-strings would probably break Node apps.
+    //
+    // As a compromise, we say:
+    // - Workers env vars that are plain strings are unchanged in process.env.
+    // - Workers env vars that can be represented as JSON will be JSON-stringified in process.env.
+    // - Anything else will be omitted.
+    //
+    // Note that you might argue that, at the config layer, it's possible to differentiate between
+    // plain strings and JSON values that evaluated to strings. Wouldn't it be nice if we could
+    // check which way the binding was originally configured in order to decide whether to
+    // represent it plain or as JSON here. However, there is no way to tell just by looking at
+    // the `env` object inside a Worker whether a particular var was originally configured as
+    // plain text, or as JSON that evaluated to a string. Either way, you just get a string. And
+    // indeed, the Workers Runtime itself does not necessarily know this. In many cases it does
+    // know, but in general the abstraction the Runtime intends to provide is that `env` is just
+    // a JavaScript object, and how exactly the contents were originally represented is not
+    // intended to be conveyed. This is important becasue, for example, we could extend dynamic
+    // dispatch bindings in the future such that the caller can specify `env` directly, and in
+    // that case the caller would simply specify a JS object, without JSON or any other
+    // serialization involved. In this case, there would be no way to know if a string var was
+    // "supposed to be" raw text vs. JSON.
+    //
+    // So, we have to do the best we can given just what we know -- the JavaScript object that is
+    // `env`.
+    //
+    // As a consolation, this is consistent with how variables are defined in wrangler.toml: you
+    // do not explicitly specify whether a variable is text or JSON. If you define a variable with
+    // a simple string value, it gets configured as a text var. If you specify an object, then it's
+    // configured as JSON.
+
+    if (typeof value === "string") {
+      env[key] = value;
+    } else if (isJsonSerializable(value)) {
+      env[key] = JSON.stringify(value);
+    }
+  }
+  return env;
+}
+
+export const env = new Proxy(getInitialEnv(), {
```

Comment:
One key difference with this approach is that the process.env will be populated lazily when first imported, correct? That means if there are manual changes to env those will be picked up, which is an inconsistency since changes to env after that point won't be picked up

Reply:
I suppose that true. Is it worth worrying about? I think it's unlikely to affect anyone, and the only way to avoid it would be to make a copy early on in C++ code which seems like a lot of work.

Reply:
Well, given that the user can introduce objects with circular references, we will, at the very least, need to address that.

Reply:
Cycles have now been addressed.